File: ios-pending/PokeEmu/PokeEmuCoreModule.m
#import <React/RCTBridgeModule.h>
#import <React/RCTEventEmitter.h>

@interface RCT_EXTERN_MODULE(PokeEmuCoreModule, RCTEventEmitter)

RCT_EXTERN_METHOD(loadROM:(NSString *)path withResolver:(RCTPromiseResolveBlock)resolve withRejecter:(RCTPromiseRejectBlock)reject)
RCT_EXTERN_METHOD(unloadROM:(RCTPromiseResolveBlock)resolve withRejecter:(RCTPromiseRejectBlock)reject)
RCT_EXTERN_METHOD(play)
RCT_EXTERN_METHOD(pause)
RCT_EXTERN_METHOD(setButtonState:(NSString *)button pressed:(BOOL)pressed)
RCT_EXTERN_METHOD(setFastForward:(BOOL)enabled speedMultiplier:(double)speedMultiplier)
RCT_EXTERN_METHOD(saveState:(NSString *)romId slotIndex:(nonnull NSNumber *)slotIndex withResolver:(RCTPromiseResolveBlock)resolve withRejecter:(RCTPromiseRejectBlock)reject)
RCT_EXTERN_METHOD(loadState:(NSString *)romId slotIndex:(nonnull NSNumber *)slotIndex withResolver:(RCTPromiseResolveBlock)resolve withRejecter:(RCTPromiseRejectBlock)reject)
RCT_EXTERN_METHOD(applyCheat:(NSString *)code enabled:(BOOL)enabled withResolver:(RCTPromiseResolveBlock)resolve withRejecter:(RCTPromiseRejectBlock)reject)
RCT_EXTERN_METHOD(removeAllCheats)

@end

File: android/app/src/main/java/com/pokeemu/core/PokeEmuCoreModule.kt
package com.pokeemu.core

import com.facebook.react.bridge.*

class PokeEmuCoreModule(reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {
  override fun getName() = "PokeEmuCore"

  @ReactMethod
  fun loadROM(path: String, promise: Promise) {
    // Task 10 replaces this stub with a real JNI call into libmgba.
    val result = Arguments.createMap()
    result.putInt("width", 240)
    result.putInt("height", 160)
    promise.resolve(result)
  }

  @ReactMethod
  fun unloadROM(promise: Promise) { promise.resolve(null) }

  @ReactMethod
  fun play() {}

  @ReactMethod
  fun pause() {}

  @ReactMethod
  fun setButtonState(button: String, pressed: Boolean) {}

  @ReactMethod
  fun setFastForward(enabled: Boolean, speedMultiplier: Double) {}

  @ReactMethod
  fun saveState(romId: String, slotIndex: Int, promise: Promise) { promise.resolve(null) }

  @ReactMethod
  fun loadState(romId: String, slotIndex: Int, promise: Promise) { promise.resolve(null) }

  @ReactMethod
  fun applyCheat(code: String, enabled: Boolean, promise: Promise) { promise.resolve(true) }

  @ReactMethod
  fun removeAllCheats() {}
}
